wire-library/wire-grpc-client/src/jvmMain/kotlin/com/squareup/wire/GrpcCalls.kt

```diff
@@ -87,3 +97,118 @@ fun <S : Any, R : Any> GrpcCall(function: (S) -> R): GrpcCall<S, R> {
     override fun clone() = GrpcCall(function)
   }
 }
+
+/**
+ * Returns a new instance of [GrpcStreamingCall] that can be used for a single call to
+ * [execute][GrpcStreamingCall.execute] or [executeBlocking][GrpcStreamingCall.executeBlocking].
+ *
+ * The returned instance launches [function] on [Dispatchers.IO]. The function must close the
+ * [SendChannel] when it has no more messages to transmit. If [function] throws, both channels will
+ * be closed using the thrown exception as a cause.
+ *
+ * This method is useful when implementing the interfaces that are generated by Wire:
+ *
+ * ```
+ *   override fun RouteChat(): GrpcStreamingCall<RouteNote, RouteNote> {
+ *     return GrpcStreamingCall { requests, responses ->
+ *       requests.consumeEach { note ->
+ *         responses.send(translateNote(note))
+ *       }
+ *       responses.close()
+ *     }
+ *   }
+ * ```
+ *
+ * It is succinct when used in an expression function:
+ *
+ * ```
+ *  override fun RouteChat() = GrpcStreamingCall<RouteNote, RouteNote> { requests, responses ->
+ *    requests.consumeEach { note ->
+ *      responses.send(translateNote(note))
+ *    }
+ *    responses.close()
+ *  }
+ * ```
+ */
+@JvmName("grpcStreamingCall")
+fun <S : Any, R : Any> GrpcStreamingCall(
+  function: suspend (ReceiveChannel<S>, SendChannel<R>) -> Unit
+): GrpcStreamingCall<S, R> {
+  return object : GrpcStreamingCall<S, R> {
+    private var canceled = AtomicBoolean()
+    private var executed = AtomicBoolean()
+    private val requestChannel = Channel<S>(1)
+    private val responseChannel = Channel<R>(1)
+
+    override val timeout: Timeout = Timeout.NONE
+
+    override fun cancel() {
+      if (canceled.compareAndSet(false, true)) {
+        requestChannel.cancel()
+        responseChannel.cancel()
+      }
+    }
+
+    override fun isCanceled() = canceled.get()
+
+    override fun isExecuted() = executed.get()
+
+    override fun execute(): Pair<SendChannel<S>, ReceiveChannel<R>> {
+      check(executed.compareAndSet(false, true)) { "already executed" }
+
+      GlobalScope.launch(Dispatchers.IO) {
```

Comment:
Is this kosher?

Reply:
Not really. The function should probably be an extension on `CoroutineScope` so that you launch into whatever scope encapsulates this functionality (and thus its cancelation signal).

This is basically `new Thread()` as it stands.

Reply:
Arg, that’s too bad.

At the point of implementing the interface there’s no coroutine scope to attach to for a lifecycle. We made `execute()` a non-suspending method, even though it returns channels.

```
 override fun RouteChat() = GrpcStreamingCall<RouteNote, RouteNote> { requests, responses ->
   requests.consumeEach { note ->
     responses.send(translateNote(note))
   }
   responses.close()
 }
```